### ISSUE TITLE
Support Storj as Decentralized object storage

### ISSUE THREAD

**Issue by solracsf** [juicedata/juicefs#1655]
**What would you like to be added**:

https://github.com/storj/storj

Storj is an S3-compatible platform and suite of decentralized applications that allows you to store data in a secure and decentralized manner. Your files are encrypted, broken into little pieces and stored in a global decentralized network of computers. Luckily, we also support allowing you (and only you) to retrieve those files!

**Go librarie available**: https://github.com/storj/uplink

**Why is this needed**:

Decentralized cloud storage represents a fundamental shift in the efficiency and economics of large-scale storage. Eliminating central control allows users to store and share data without reliance on a third-party storage provider. Decentralization mitigates the risk of data failures and outages while simultaneously increasing the security and privacy of object storage.

Please note that a S3-gateway is available, but this opts-in for a server-side encryption; **using Uplink**, end-to-end encryption is available and performance is superior.

**Comment by davies** [comment]
It looks like uplink is still in early phase, we would like to keep eye on it, don't depend on it for now.

**Comment by solracsf** [comment]
Hi, what do you call "early phase"? 🤔

**Comment by amwolff** [comment]
Of course, it depends on how one defines an early phase, but https://github.com/storj/uplink is a production-grade library at the moment. The S3-compatible object storage from Storj that JuiceFS can use as its backend uses uplink internally to communicate with the network.

**Comment by solracsf** [comment]
The main advantage of using Uplink library instead of S3 Gateway...is to avoid using the gateway :-) and introduce a SPOF.

**Comment by seth-hunter** [comment]
+1, very desirable feature. The ListObjects limitation of the MT gateway is problematic and the ST gateway workaround is resource intensive, slow, and seemingly buggy. Native uplink library interface seems promising.

**Comment by eleaner** [comment]
+1, fully agree

**Comment by hannahkamundson** [comment]
I would like to work on this!

**Comment by solracsf** [comment]
> It looks like uplink is still in early phase, we would like to keep eye on it, don't depend on it for now.

Hi @davies, do you believe this could now be considered?

**Comment by davies** [comment]
We can accept PR for uplink.
